src/sentry/eventstream/kafka/consumer.py

```diff
@@ -0,0 +1,115 @@
+from __future__ import absolute_import
+
+import logging
+import uuid
+
+from confluent_kafka import Consumer, OFFSET_BEGINNING, TopicPartition
+
+from sentry.eventstream.kafka.state import PartitionState, SynchronizedPartitionStateManager
+from sentry.eventstream.kafka.utils import join
+
+
+logger = logging.getLogger(__name__)
+
+
+class SynchronizedConsumer(object):
+    def __init__(self, bootstrap_servers, topics, consumer_group, commit_log_topic, synchronize_commit_group):
+        self.bootstrap_servers = bootstrap_servers
+        self.topics = topics
+        self.consumer_group = consumer_group
+        self.commit_log_topic = commit_log_topic
+        self.synchronize_commit_group = synchronize_commit_group
+
+    def _commit_callback(self, error, partitions):
+        if error is not None:
+            logger.warning('Failed to commit offsets (error: %s, partitions: %r)', error, partitions)
+
+    def _on_partition_state_change(self, topic, partition, previous_state_and_offsets, current_state_and_offsets):
+        logger.debug('State change for %r: %r to %r', (topic, partition), previous_state_and_offsets, current_state_and_offsets)
+
+        current_state, current_offsets = current_state_and_offsets
+        if current_offsets.local is None:
+            return  # it only makes sense to manipulate the consumer if we've got an assignment
+
+        if current_state in (PartitionState.UNKNOWN, PartitionState.SYNCHRONIZED, PartitionState.REMOTE_BEHIND):
+            self._consumer.pause([TopicPartition(topic, partition, current_offsets.local)])
+        elif current_state is PartitionState.LOCAL_BEHIND:
+            self._consumer.resume([TopicPartition(topic, partition, current_offsets.local)])
+        else:
+            raise NotImplementedError('Unexpected partition state: %s' % (current_state,))
+
+    def _get_initial_offset(self, consumer, i):
+        low, high = self._consumer.get_watermark_offsets(i)
+        return low
+
+    def _pause_partitions_on_assignment(self, consumer, assignment):
+        assignment = [
+            TopicPartition(
+                i.topic,
+                i.partition,
+                i.offset if i.offset > -1 else self._get_initial_offset(consumer, i),
+            ) for i in assignment
+        ]
+
+        self._consumer.assign(assignment)
```

Comment:
I need to verify this actually picks up the committed offsets correctly and doesn't just go to the default initial offset (low watermark.) This is one spot that would really benefit from an integration test